### PR TITLE
Switch to official Rocky image

### PR DESCRIPTION
**Pull request by paddyroddy**
Think it's better to use an official image. Also, aligns with SRR.

### CODE REVIEW

File: molecule_configs/rocky9_base_config.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ driver:
 platforms:
   - name: instance
     hostname: molecule.instance.local
-    image: ${MOLECULE_DOCKER_IMAGE:-geerlingguy/docker-rockylinux9-ansible:latest}
+    image: ${MOLECULE_DOCKER_IMAGE:-rockylinux/rockylinux:9-ubi-init}
```
Comment (drmatthews) [inline molecule_configs/rocky9_base_config.yml:15]:
```suggestion
    image: rockylinux/rockylinux:9-ubi-init
```

File: playbooks/molecule/rocky9_monitoring/molecule.yml [hunk 2]
```diff
@@ -12,7 +12,7 @@ driver:
 platforms:
   - name: mclient
     hostname: molecule.mclient.local
-    image: ${MOLECULE_DOCKER_IMAGE:-geerlingguy/docker-rockylinux9-ansible:latest}
+    image: ${MOLECULE_DOCKER_IMAGE:-rockylinux/rockylinux:9-ubi-init}
```
Comment (drmatthews) [inline playbooks/molecule/rocky9_monitoring/molecule.yml:15]:
```suggestion
    image: rockylinux/rockylinux:9-ubi-init
```

File: roles/nginx/molecule/rocky9/molecule.yml [hunk 3]
```diff
@@ -4,7 +4,7 @@
 platforms:
   - name: instance
     hostname: molecule.instance.local
-    image: ${MOLECULE_DOCKER_IMAGE:-geerlingguy/docker-rockylinux9-ansible:latest}
+    image: ${MOLECULE_DOCKER_IMAGE:-rockylinux/rockylinux:9-ubi-init}
```
Comment (drmatthews) [inline roles/nginx/molecule/rocky9/molecule.yml:7]:
```suggestion
    image: rockylinux/rockylinux:9-ubi-init
```

**Review by drmatthews — CHANGES_REQUESTED** [review]
Looks good. I removed remaining occurrences of `MOLECULE_DOCKER_IMAGE` for consistency.

**Review by p-j-smith — CHANGES_REQUESTED** [review]
looks good! Although there are a few more roles that need the prepare steps added:
- [X] firewalld
- [X] postgresql
- [X] provision
- [X] provision_accounts

Side note (and can be done in a different PR), but the workflows for testing roles should run when the molecule base configs change (i.e. add `- "molecule_configs/*"` to the paths)

**Comment by paddyroddy** [comment]
> looks good! Although there are a few more roles that need the prepare steps added:
> 
>     * [x]  firewalld
> 
>     * [x]  postgresql
> 
>     * [x]  provision
> 
>     * [x]  provision_accounts
> 
> 
> Side note (and can be done in a different PR), but the workflows for testing roles should run when the molecule base configs change (i.e. add `- "molecule_configs/*"` to the paths)

I've done all of this but can see it's failing for `sudo` missing reasons, any ideas?

**Comment by p-j-smith** [comment]
I've done all of this but can see it's failing for sudo missing reasons, any ideas?

I haven't looked at monitoring, but for postgresql the sudo installation is happening after the `mirsg.infrastructure.provision` role runs:
https://github.com/UCL-MIRSG/ansible-collection-infra/blob/e8cbd31ec234296e12d0a0900e7b8c0533659c8b/roles/postgresql/molecule/resources/prepare.yml#L7C1-L11C23
You'll need to add a pre-task to install sudo so it runs before the role:
```yaml
  pre_tasks:
    - name: Install sudo
      ansible.builtin.package:
        name: sudo
        state: present
```

**Comment by p-j-smith** [comment]
For the monitoring there's a [separate prepare playbook](https://github.com/UCL-MIRSG/ansible-collection-infra/blob/main/playbooks/molecule/resources/monitoring/prepare.yml) that needs the sudo installation added

**Comment by paddyroddy** [comment]
> For the monitoring there's a [separate prepare playbook](https://github.com/UCL-MIRSG/ansible-collection-infra/blob/main/playbooks/molecule/resources/monitoring/prepare.yml?rgh-link-date=2024-01-23T15%3A12%3A52Z) that needs the sudo installation added

Of course, I should have known!

**Comment by paddyroddy** [comment]
Now getting for `molecule-monitoring (rocky9_monitoring)`
image

**Comment by p-j-smith** [comment]
It looks like the issue is that the SSL certificates are being generated before cryptograph is installed:
https://github.com/UCL-MIRSG/ansible-collection-infra/blob/2e26c0393255179a27a76ed865db48850694b194/playbooks/molecule/resources/monitoring/converge.yml#L7-L12
I think you'll need to add the role `mirsg.infrastructure.install_python` before `mirsg.infrastructure.ssl_certificates`

**Comment by paddyroddy** [comment]
> I think you'll need to add the role `mirsg.infrastructure.install_python` before `mirsg.infrastructure.ssl_certificates`

This was the error this time
image

I will try pulling in your latest changes

**Comment by paddyroddy** [comment]
I'm going to need help
image

**Comment by paddyroddy** [comment]
Would either of you @drmatthews @p-j-smith be able to look into this? I don't know `molecule`, so I'm a bit lost. This is my [XNAT Dockerfile used in SRR](https://github.com/UCL-MIRSG/UCLH-MPBE-SRR-XNAT/blob/main/docker/xnat.Dockerfile)

**Comment by drmatthews** [comment]
Sorry for the lack of a response @paddyroddy, been in meetings all day. Will try to look tomorrow.

**Comment by paddyroddy** [comment]
Anyone had a chance to look?

**Comment by drmatthews** [comment]
Apologies @paddyroddy - I've been tied up with a few other things. It looks like a `cgroup` thing but I need to look in more detail. The monitoring consists of a set of Docker containers running on the monitoring host. Some of them require `cgroup` volumes to be mounted (in the tests the monitoring host is itself a Docker container). There isn't a lot going on in the [Docker file we're using](https://github.com/geerlingguy/docker-rockylinux9-ansible/blob/master/Dockerfile) but it does define this `VOLUME`:

```
VOLUME ["/sys/fs/cgroup"]
```

But the purpose of that is just make sure that the volume is always mounted when a container is created. This volume is explicitly set in `rocky9_base_config.yml`:

https://github.com/UCL-MIRSG/ansible-collection-infra/blob/f39813a234b5fc1dc3e995c9c57b12901bb70871/molecule_configs/rocky9_base_config.yml#L21-L23

So it is unclear why it isn't working. I'll need to find some time to test some options for the volumes being mounted in the `cadvisor` container (this is throwing the error) and some options for the modifying the Docker file for the image used by molecule.

**Comment by paddyroddy** [comment]
I'm using the same `cgroup` thing in SRR https://github.com/UCL-MIRSG/UCLH-MPBE-SRR-XNAT/blob/1cb6f5e1e3e38db61827d7ff3ad24f3fac0f17a3/docker-compose.yml#L65, and had spotted it in your config so didn't think that was causing the issue.

**Comment by drmatthews** [comment]
The error says `error mounting \"/sys/fs/cgroup/cpu,cpuacct\" to rootfs`. This a volume being mounted into the `cadvisor` container from the molecule `mserv` container.

**Comment by drmatthews** [comment]
> The error says error mounting \"/sys/fs/cgroup/cpu,cpuacct\" to rootfs. This a volume being mounted into the cadvisor container from the molecule mserv container.

Turns out I had conflicting volumes defined for the `cadvisor` container. Don't know why it wasn't picked up before.

**Review by drmatthews — APPROVED** [review]
Looks good - sorry for the wait.

**Review by p-j-smith — APPROVED** [review]
nice 🎉

**Pull request merged**